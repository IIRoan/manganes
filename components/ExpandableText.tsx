import React, { useState, useCallback, useMemo } from 'react';
import { View, Text, TouchableOpacity, StyleSheet, TextStyle, Platform, TextInput, LayoutAnimation } from 'react-native';

interface ExpandableTextProps {
  text: string;
  initialLines?: number;
  style?: TextStyle;
  expandedStyle?: TextStyle;
}

const ExpandableText: React.FC<ExpandableTextProps> = ({ 
  text, 
  initialLines = 3, 
  style,
  expandedStyle,
}) => {
  const [isExpanded, setIsExpanded] = useState(false);
  const [isTruncated, setIsTruncated] = useState(false);

  const onTextLayout = useCallback((e: any) => {
    if (e.nativeEvent.lines.length > initialLines) {
      setIsTruncated(true);
    }
  }, [initialLines]);

  const toggleExpand = useCallback(() => {
    LayoutAnimation.configureNext(LayoutAnimation.Presets.easeInEaseOut);
    setIsExpanded((prev) => !prev);
  }, []);

  const TextComponent = useMemo(() => {
    return Text;
  }, [text]);

  const textProps = useMemo(() => {
    if (TextComponent === TextInput) {
      return {
        multiline: true,
        editable: false,
        scrollEnabled: false,
      };
    }
    return {
      numberOfLines: isExpanded ? undefined : initialLines,
      onTextLayout: onTextLayout,
    };
  }, [TextComponent, isExpanded, initialLines, onTextLayout]);

  return (
    <TouchableOpacity 
      onPress={toggleExpand}
      activeOpacity={0.7}
      style={styles.container}
    >
      <TextComponent
        {...textProps}
        style={[
          styles.text, 
          style,
          isExpanded && expandedStyle,
          isTruncated && !isExpanded && styles.truncatedText
        ]}
      >
        {text}
      </TextComponent>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  container: {
    overflow: 'hidden',
  },
  text: {
    fontSize: 16,
    lineHeight: 24,
  },
  truncatedText: {
    marginBottom: 4, // Add some space for the fade effect
  },
});

export default React.memo(ExpandableText);
