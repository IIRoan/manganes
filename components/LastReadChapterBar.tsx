import React, { useEffect } from 'react';
import { View, Text, TouchableOpacity, StyleSheet } from 'react-native';
import { Ionicons } from '@expo/vector-icons';

interface LastReadChapterBarProps {
  lastReadChapter: string | null;
  onPress: () => void;
  colors: any;
  readChapters: string[];
}
const LastReadChapterBar: React.FC<LastReadChapterBarProps> = ({ 
  lastReadChapter, 
  onPress, 
  colors 
}) => {
  const isStartReading = !lastReadChapter || lastReadChapter === 'Not started';

  return (
    <TouchableOpacity 
      onPress={onPress} 
      style={[styles.container, { backgroundColor: colors.background }]}
      testID="last-read-chapter-bar"
    >
      <View style={styles.content}>
        <Ionicons 
          name={isStartReading ? "play-circle-outline" : "bookmark"} 
          size={20} 
          color={colors.primary} 
        />
        <Text style={[styles.text, { color: colors.text }]}>
          {isStartReading ? (
            "Start reading"
          ) : (
            <>
              Continue from <Text style={[styles.chapterText, { color: colors.primary }]}>
                {lastReadChapter}
              </Text>
            </>
          )}
        </Text>
      </View>
      <Ionicons name="chevron-forward" size={20} color={colors.primary} />
    </TouchableOpacity>
  );
};


const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    borderRadius: 10,
    marginTop: 20,
    marginBottom: -20,
  },
  content: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  text: {
    marginLeft: 10,
    fontSize: 14,
  },
  chapterText: {
    fontWeight: 'bold',
  },
});

export default LastReadChapterBar;
