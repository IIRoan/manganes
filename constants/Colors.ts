// constants/Colors.ts

export const Colors = {
  light: {
    primary: '#2E8B57',
    background: '#F5F5F5',
    card: '#FFFFFF',
    text: '#333333',
    border: '#E0E0E0',
    notification: '#FF6B6B',
    tint: '#2E8B57',
    tabIconDefault: '#BDBDBD',
    tabIconSelected: '#2E8B57',
    secondaryText: '#757575',
    error: '#FF6B6B',

  },
  dark: {
    primary: '#4CAF50',
    background: '#121212',
    card: '#1f1f1d',
    text: '#E0E0E0',
    border: '#333333',
    notification: '#FF6B6B',
    tint: '#4CAF50',
    tabIconDefault: '#757575',
    tabIconSelected: '#4CAF50',
    secondaryText: '#BDBDBD',
    error: '#FF6B6B',
  },
};

export type ColorScheme = keyof typeof Colors;
export type ThemeColors = typeof Colors.light & typeof Colors.dark;
