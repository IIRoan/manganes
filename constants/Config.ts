export const MANGA_API_URL = 'https://mangafire.to';
