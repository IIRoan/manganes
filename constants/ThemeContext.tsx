import React, { createContext, useState, useContext, useEffect } from 'react';
import { useColorScheme } from 'react-native';
import { ColorScheme } from '@/constants/Colors';

type Theme = 'light' | 'dark' | 'system';

interface ThemeContextType {
  theme: Theme;
  systemTheme: ColorScheme;
  setTheme: (theme: Theme) => void;
  toggleTheme: () => void;
  actualTheme: 'light' | 'dark';
}

const ThemeContext = createContext<ThemeContextType | undefined>(undefined);

export const ThemeProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const [theme, setTheme] = useState<Theme>('system');
  const systemColorScheme = useColorScheme() as ColorScheme;

  const toggleTheme = () => {
    setTheme(prevTheme => {
      switch (prevTheme) {
        case 'light': return 'dark';
        case 'dark': return 'system';
        case 'system': return 'light';
      }
    });
  };

  const actualTheme = theme === 'system' ? systemColorScheme : theme;

  return (
    <ThemeContext.Provider value={{ 
      theme, 
      systemTheme: systemColorScheme, 
      setTheme,
      toggleTheme,
      actualTheme
    }}>
      {children}
    </ThemeContext.Provider>
  );
};

export const useTheme = () => {
  const context = useContext(ThemeContext);
  if (context === undefined) {
    throw new Error('useTheme must be used within a ThemeProvider');
  }
  return context;
};