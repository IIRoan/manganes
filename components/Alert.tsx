import React from 'react';
import { View, Text, TouchableOpacity, StyleSheet, Modal, useColorScheme, TouchableWithoutFeedback, Animated } from 'react-native';
import { Colors, ColorScheme } from '@/constants/Colors';
import { useTheme } from '@/constants/ThemeContext';
import { Ionicons } from '@expo/vector-icons';

type IconName = 'options' | 'key' | 'search' | 'repeat' | 'link' | 'at' | 'push' | 'map' | 'filter' | 'scale' | 'body' | 'code' | 'menu' | 'time' | 'ellipse' | 'image' | 'stop' | 'text' | 'alert' | 'book-outline' | 'book' | 'checkmark-circle-outline' | 'close-circle-outline';

interface Option {
  text: string;
  onPress: () => void;
  icon?: IconName;
}

interface CustomAlertProps {
  visible: boolean;
  title: string;
  onClose: () => void;
  type: 'bookmarks' | 'confirm';
  options?: Option[];
  message?: string;
}

const Alert: React.FC<CustomAlertProps> = ({ visible, title, onClose, type, options, message }) => {
  const { theme } = useTheme();
  const systemColorScheme = useColorScheme() as ColorScheme;
  const colorScheme = theme === 'system' ? systemColorScheme : theme as ColorScheme;
  const colors = Colors[colorScheme];

  const styles = getStyles(colors);
  const scaleValue = React.useRef(new Animated.Value(0)).current;

  React.useEffect(() => {
    if (visible) {
      Animated.spring(scaleValue, {
        toValue: 1,
        useNativeDriver: true,
      }).start();
    } else {
      Animated.spring(scaleValue, {
        toValue: 0,
        useNativeDriver: true,
      }).start();
    }
  }, [visible]);

  const renderContent = () => {
    if (type === 'bookmarks') {
      return (
        <View testID="alert-options" style={styles.optionsContainer}>
          {options?.map((option, index) => (
            <TouchableOpacity
              key={index}
              style={styles.button}
              onPress={() => {
                option.onPress();
                onClose();
              }}
            >
              {option.icon && (
                <View style={styles.iconContainer}>
                  <Ionicons name={option.icon} size={24} color={colors.primary} />
                </View>
              )}
              <Text style={styles.textStyle}>{option.text}</Text>
            </TouchableOpacity>
          ))}
        </View>
      );
    } else if (type === 'confirm') {
      return (
        <>
          <Text style={styles.message}>{message}</Text>
          <View style={styles.confirmButtonsContainer}>
            {options?.map((option, index) => (
              <TouchableOpacity
                key={index}
                style={styles.confirmButton}
                onPress={() => {
                  option.onPress();
                  onClose();
                }}
              >
                <Text style={styles.confirmButtonText}>{option.text}</Text>
              </TouchableOpacity>
            ))}
          </View>
        </>
      );
    }
  };

  return (
    <Modal
      transparent={true}
      visible={visible}
      onRequestClose={onClose}
      animationType="fade"
    >
      <TouchableWithoutFeedback onPress={onClose}>
        <View style={styles.centeredView}>
          <TouchableWithoutFeedback>
            <Animated.View style={[styles.modalView, { transform: [{ scale: scaleValue }] }]}>
              <TouchableOpacity testID="close-button" style={styles.closeButton} onPress={onClose}>
                <Ionicons name="close" size={24} color={colors.text} />
              </TouchableOpacity>
              <Text testID="alert-title" style={styles.modalText}>{title}</Text>
              {renderContent()}
            </Animated.View>
          </TouchableWithoutFeedback>
        </View>
      </TouchableWithoutFeedback>
    </Modal>
  );
};

const getStyles = (colors: typeof Colors.light) => StyleSheet.create({
  centeredView: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'rgba(0, 0, 0, 0.5)',
  },
  modalView: {
    margin: 20,
    backgroundColor: colors.card,
    borderRadius: 24,
    padding: 24,
    alignItems: 'stretch',
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 4,
    },
    shadowOpacity: 0.25,
    shadowRadius: 8,
    elevation: 5,
    minWidth: 300,
    maxWidth: '85%',
  },
  optionsContainer: {
    marginTop: 16,
  },
  button: {
    flexDirection: 'row',
    alignItems: 'center',
    borderRadius: 12,
    padding: 14,
    marginTop: 12,
    backgroundColor: colors.card,
    borderWidth: 1,
    borderColor: colors.border,
  },
  iconContainer: {
    width: 40,
    height: 40,
    borderRadius: 20,
    backgroundColor: `${colors.primary}15`,
    justifyContent: 'center',
    alignItems: 'center',
    marginRight: 16,
  },
  textStyle: {
    color: colors.text,
    fontWeight: '600',
    fontSize: 16,
    flex: 1,
  },
  modalText: {
    marginBottom: 16,
    textAlign: 'center',
    fontWeight: 'bold',
    fontSize: 24,
    color: colors.primary,
  },
  closeButton: {
    position: 'absolute',
    right: 16,
    top: 16,
    zIndex: 1,
    width: 32,
    height: 32,
    borderRadius: 16,
    backgroundColor: `${colors.text}10`,
    justifyContent: 'center',
    alignItems: 'center',
  },
  message: {
    fontSize: 16,
    color: colors.text,
    textAlign: 'center',
    marginBottom: 20,
  },
  confirmButtonsContainer: {
    flexDirection: 'row',
    justifyContent: 'space-around',
  },
  confirmButton: {
    backgroundColor: colors.primary,
    paddingVertical: 10,
    paddingHorizontal: 20,
    borderRadius: 8,
  },
  confirmButtonText: {
    color: colors.card,
    fontSize: 16,
    fontWeight: 'bold',
  },
});

export default Alert;