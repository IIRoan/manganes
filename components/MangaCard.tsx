import React from 'react';
import { View, Text, Image, TouchableOpacity, StyleSheet } from 'react-native';
import { Colors, ColorScheme } from '@/constants/Colors';
import { useTheme } from '@/constants/ThemeContext';

interface MangaCardProps {
  title: string;
  imageUrl: string;
  onPress: () => void;
}

const MangaCard: React.FC<MangaCardProps> = ({ title, imageUrl, onPress }) => {
  const { theme, systemTheme } = useTheme();
  const colorScheme = theme === 'system' ? systemTheme : theme as ColorScheme;
  const colors = Colors[colorScheme];

  return (
    <TouchableOpacity 
      onPress={onPress} 
      style={[
        styles.card, 
        { 
          backgroundColor: colors.card,
          borderColor: colors.border,
        }
      ]}
    >
      <Image source={{ uri: imageUrl }} style={styles.image} />
      <View style={styles.titleContainer}>
        <Text 
          style={[
            styles.title, 
            { color: colors.text }
          ]} 
          numberOfLines={2}
        >
          {title}
        </Text>
      </View>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  card: {
    width: '48%',
    marginBottom: 10,
    borderRadius: 8,
    overflow: 'hidden',
    borderWidth: 1,
  },
  image: {
    width: '100%',
    height: 200,
    resizeMode: 'cover',
  },
  titleContainer: {
    padding: 8,
  },
  title: {
    fontSize: 14,
    fontWeight: 'bold',
  },
});

export default MangaCard;