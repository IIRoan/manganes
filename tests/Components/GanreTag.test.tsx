import React from 'react';
import { render } from '@testing-library/react-native';
import { GenreTag } from '@/components/GanreTag';
import { useTheme } from '@/constants/ThemeContext';

jest.mock('@/constants/ThemeContext', () => ({
  useTheme: jest.fn(),
}));

describe('GenreTag Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.mocked(useTheme).mockReturnValue({
      theme: 'light',
      systemTheme: 'light',
      setTheme: jest.fn(),
      toggleTheme: jest.fn(),
      actualTheme: 'light',
    });
  });

  it('renders the genre text', () => {
    const { getByText } = render(<GenreTag genre="Action" />);
    expect(getByText('Action')).toBeTruthy();
  });

  it('matches snapshot', () => {
    const tree = render(<GenreTag genre="Drama" />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
