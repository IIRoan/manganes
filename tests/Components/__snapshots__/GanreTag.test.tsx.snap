// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenreTag Component matches snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "#2E8B57",
      "borderRadius": 15,
      "margin": 2,
      "paddingHorizontal": 10,
      "paddingVertical": 5,
    }
  }
>
  <Text
    style={
      {
        "color": "#FFFFFF",
        "fontSize": 12,
        "fontWeight": "600",
      }
    }
  >
    Drama
  </Text>
</View>
`;
